src/Symfony/Component/Cache/Traits/RedisTrait.php

```diff
@@ -118,8 +118,17 @@ public static function createConnection($dsn, array $options = array())
         $params += $options + self::$defaultConnectionOptions;
         if (null === $params['class'] && !\extension_loaded('redis') && !class_exists(\Predis\Client::class)) {
             throw new CacheException(sprintf('Cannot find the "redis" extension, and "predis/predis" is not installed: %s', $dsn));
+        } elseif (isset($params['predis_options']) && !class_exists(\Predis\Client::class)) {
+            throw new CacheException(sprintf('DSN contains "predis_options" and "predis/predis" is not installed: %s', $dsn));
+        } elseif (isset($params['predis_options']) || !\extension_loaded('redis')) {
+            if (null !== $params['class'] && !is_a($params['class'], \Predis\Client::class, true)) {
+                throw new InvalidArgumentException(sprintf('Cannot find the "redis" extension and expected an instance of %s, instead got: %s', \Predis\Client::class, $params['class']));
```

Comment:
the message is confusing: when the extension *is* installed, we should tell otherwise.